src/MLIAP/mliap_descriptor_so3.cpp

```diff
@@ -0,0 +1,298 @@
+/* ----------------------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://lammps.sandia.gov/, Sandia National Laboratories
+   Steve Plimpton, sjplimp@sandia.gov
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+#include "mliap_descriptor_so3.h"
+#include "atom.h"
+#include "comm.h"
+#include "error.h"
+#include "memory.h"
+#include "mliap_data.h"
+#include "pair_mliap.h"
+#include "mliap_so3.h"
+
+#include <cmath>
+#include <cstring>
+
+using namespace LAMMPS_NS;
+
+#define MAXLINE 1024
+#define MAXWORD 3
+
+
+MLIAPDescriptorSO3::MLIAPDescriptorSO3(LAMMPS *lmp,
+                                       char *paramfilename):
+  MLIAPDescriptor(lmp)
+{
+  nelements = 0;
+  elements = nullptr;
+  radelem = nullptr;
+  wjelem = nullptr;
+  so3ptr = nullptr;
+  read_paramfile(paramfilename);
+
+  so3ptr = new MLIAP_SO3(lmp,rcutfac,lmax,nmax,alpha);
+
+  ndescriptors = so3ptr->ncoeff;
+}
+
+MLIAPDescriptorSO3::~MLIAPDescriptorSO3()
+{
+  delete so3ptr;
+}
+
+void MLIAPDescriptorSO3::read_paramfile(char *paramfilename)
+{
+  int rcutfacflag = 0;
+  int nelementsflag = 0;
+  int elementsflag = 0;
+  int radelemflag = 0;
+  int wjelemflag = 0;
+  int nmaxflag=0;
+  int lmaxflag=0;
+  int alphaflag=0;
+
+  // set defaults for optional keywords
+
+  rfac0 = 0.99363;
+  rmin0 = 0.0;
+
+  // open SO3 parameter file on proc 0
+
+  FILE *fpparam;
+  if (comm->me == 0) {
+    fpparam = utils::open_potential(paramfilename,lmp,nullptr);
+    if (fpparam == nullptr)
+      error->one(FLERR,fmt::format
+        ("Cannot open SO3 parameter file {}: {}",
+          paramfilename, utils::getsyserror()));
+  }
+
+  char line[MAXLINE],*ptr;
+  int eof = 0;
+  int n,nwords;
+
+  while (1) {
```

Comment:
We are trying to get rid of this kind of file parser code construct and reduce the use of C-style string functions and particularly get rid of `strtok()` (which is not thread safe). There are multiple convenience classes and functions available. Please see https://docs.lammps.org/Developer_utils.html and the examples given or how those are used throughout other places in LAMMPS. This is a work in progress, but that is why we are particularly interested in having newly added code adopt these tools.